### ISSUE TITLE
[Form] Grouping entity choices produces PHP warning when grouping by another entity

### ISSUE THREAD

**Issue by adamelso** [symfony/symfony#14976]
I'm not certain if this is a bug or not, but one breaking change I discovered between 2.6 and 2.7 is when you add a form entity choice type and attempt to group the choices by a related entity, this will produce a PHP warning and ultimately a `ContextErrorException` because it uses the entity object as an array key.

> Warning: Illegal offset type in isset or empty in [...] DefaultChoiceListFactory.php on line 334

In the example below, I'm attempting to group `province` choices - instances of `Province` by its `country` property, an instance of `Country`.

```php
$builder->add('province', 'entity', [
                'choices' => $provinces,
                'class' => Province::class,
                'group_by' => 'country',
                'empty_value' => 'All Provinces',
                'empty_data'  => null,
                'required' => false,
                // @todo check if Symfony 2.7 docs needs updating for the options below.
                'property' => 'name',
                'choice_label' => 'name',
]);
```

It works on 2.6, but breaks in 2.7.

I was able to fix this by defining a more explicit path for the `group_by` option.

    'group_by' => 'country.name',

I was also able to fix this by modifying `DefaultChoiceListFactory:: addChoiceViewGroupedBy()` to cast the $groupLabel callable to a string, (but the former solution is better :).

```php
private static function addChoiceViewGroupedBy($groupBy, $choice, $key, $label, $values, &$index, $attr, $isPreferred, &$preferredViews, &$otherViews)
{
    $groupLabel = (string) call_user_func($groupBy, $choice, $key, $values[$key]);
    //...
}
```

So I have 3 questions:

1. Was grouping by objects always supported?
2. Should Symfony Form cast to a string as a fallback?
3. Should a more helpful exception occur otherwise?

**Comment by webmozart** [comment]
Hi @adamelso, thank you for reporting this issue! Casting to a string sounds like a good solution. See #15064 for the PR.

**Comment by klepsydra** [comment]
This has been plaguing my project too, and I didn't even know where to search.
http://strat.sqr-dev.pl/app_dev.php/user/strategies/new
```
 ->add('preset', null, [
'property' => 'SelectionDisplayName',
'multiple' => FALSE,
'class' => 'StrategioMainBundle:Presets', 
'group_by' => 'user',  # Used to worked in Symfony 2.6
])

```


Now with Symfony 2.7.1:

`
 Warning: Illegal offset type in isset or empty
500 Internal Server Error - ContextErrorException 


Stack Trace

    in vendor/symfony/symfony/src/Symfony/Component/Form/ChoiceList/Factory/DefaultChoiceListFactory.php at line 334  -
                // Initialize the group views if necessary. Unnnecessarily built group
                // views will be cleaned up at the end of createView()
                if (!isset($preferredViews[$groupLabel])) {                   # line 334  
                    $preferredViews[$groupLabel] = new ChoiceGroupView($groupLabel);
                    $otherViews[$groupLabel] = new ChoiceGroupView($groupLabel);
                }
`




However, I still can't get it to work as before, though I tried various:
'class' => \Strategio\MainBundle\Entity\Presets::class, # no difference from 'StrategioMainBundle:Presets'

                    'group_by' => 'Presets.getUser()', 
                    'group_by' => 'presets.user', 

'
 Neither the property "Presets" nor one of the methods "getPresets()", "presets()", "isPresets()", "hasPresets()", "__get()" exist and have public access in class "Strategio\MainBundle\Entity\Presets". 

'

**Comment by webmozart** [comment]
Fixed in #15064.
